internal/configs/action.go

```diff
@@ -0,0 +1,275 @@
+	DependsOn []hcl.Traversal
+
+	ProviderConfigRef *ProviderConfigRef
+	Provider          addrs.Provider
+
+	DeclRange hcl.Range
+	TypeRange hcl.Range
+}
+
+// ActionTrigger represents a configured "action_trigger" inside the lifecycle
+// block of a managed resource.
+type ActionTrigger struct {
+	Condition hcl.Expression
+	Events    []ActionTriggerEvent
+	Actions   []ActionRef
+
+	DeclRange hcl.Range
+}
+
+// ActionTriggerEvent is an enum for valid values for events for action
+// triggers.
+type ActionTriggerEvent int
+
+//go:generate go tool golang.org/x/tools/cmd/stringer -type ActionTriggerEvent
+
+const (
+	BeforeCreate ActionTriggerEvent = iota
+	AfterCreate
+	BeforeUpdate
+	AfterUpdate
+	BeforeDestroy
+	AfterDestroy
+)
+
+// ActionRef represents a reference to a configured Action
+// copypasta of providerconfigref; not sure what's needed.
+type ActionRef struct {
+	Traversal hcl.Traversal
+	Range     hcl.Range
+}
+
+func decodeActionTriggerBlock(block *hcl.Block) (*ActionTrigger, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	a := &ActionTrigger{
+		Events:    []ActionTriggerEvent{},
+		Actions:   []ActionRef{},
+		Condition: nil,
+	}
+
+	content, bodyDiags := block.Body.Content(actionTriggerSchema)
+	diags = append(diags, bodyDiags...)
+
+	if attr, exists := content.Attributes["condition"]; exists {
+		a.Condition = attr.Expr
+	}
+
+	// this is parsing events like expressions, so it's angry when there's quotes
+	// needs to parse strings:
+	// Quoted references are deprecated; In this context, references are expected literally rather than in quotes.
+	if attr, exists := content.Attributes["events"]; exists {
+		exprs, ediags := hcl.ExprList(attr.Expr)
+		diags = append(diags, ediags...)
+
+		events := []ActionTriggerEvent{}
+
+		for _, expr := range exprs {
+
+			expr, shimDiags := shimTraversalInString(expr, true)
+			diags = append(diags, shimDiags...)
+
+			var event ActionTriggerEvent
+			switch hcl.ExprAsKeyword(expr) {
+			case "before_create":
+				event = BeforeCreate
+			case "after_create":
+				event = AfterCreate
+			case "before_update":
+				event = BeforeUpdate
+			case "after_update":
+				event = AfterUpdate
+			case "before_destroy":
+				event = BeforeDestroy
+			case "after_destroy":
+				event = AfterDestroy
+			default:
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  fmt.Sprintf("Invalid \"event\" value %s", hcl.ExprAsKeyword(expr)),
+					Detail:   "The \"event\" argument supports the following values: before_create, after_create, before_update, after_update, before_destroy, after_destroy.",
+					Subject:  expr.Range().Ptr(),
+				})
+			}
+			events = append(events, event)
+		}
+		a.Events = events
+	}
+
+	if attr, exists := content.Attributes["actions"]; exists {
+		exprs, ediags := hcl.ExprList(attr.Expr)
+		diags = append(diags, ediags...)
+		actions := []ActionRef{}
+		for _, expr := range exprs {
+			expr, shimDiags := shimTraversalInString(expr, false)
```

Comment:
`shimTraversalInString` is for compatibility with <0.12 config, so should not be used in any new features

Reply:
removed, ty 👍🏻